### PR TITLE
Fix svg size decoder

### PR DESCRIPTION
**Pull request by thierry-f-78**
Hi, this PR contains contains some improvement to handle SVG as path source for PDF:
- fix a bug about units in width and height of th SVG
- support basic rect form
- support drawing forms with DrawPath() (to fill the form with a color)

### CODE REVIEW

**Review by sbinet — CHANGES_REQUESTED** [review]
thanks for the PR.

could you address the comments from `staticcheck` ?

```
Error: svgbasic.go:230:52: 'bitSize' argument is invalid, must be either 32 or 64 (SA1030)
Error: svgbasic.go:306:4: this value of err is never used (SA4006)
Error: svgbasic.go:306:10: Errorf doesn't have side effects and its return value is ignored (SA4017)
```

also, could you perhaps add a test or example for this new API ?

thanks again.

**Comment by thierry-f-78** [comment]
Hi, static check is fixed. I wrote examples asap.

**Comment by thierry-f-78** [comment]
example done.

File: svgbasic.go [hunk 1]
```diff
@@ -200,36 +201,113 @@ type SVGBasicType struct {
 	Segments [][]SVGBasicSegmentType
 }
 
+// This function parse float and its unit. The result is converted in pt
+// wich is the default document value. The function return the scale
+// to apply to each position or distance value.
```
Comment (sbinet) [inline svgbasic.go:206]:
```suggestion
// parseFloatWithUnit parses a float and its unit, e.g. "42pt".
//
// The result is converted into pt values wich is the default document unit.
// parseFloatWithUnit returns the factor to apply to positions or distances to
// convert their values in point units.
```

File: svgbasic.go [hunk 2]
```diff
@@ -200,36 +201,113 @@ type SVGBasicType struct {
...
+	} else if strings.HasSuffix(val, "cm") {
+		removeUnitChar = 2
+		adjustToPt = 72.0 / 2.54
+	} else if strings.HasSuffix(val, "pc") {
+		removeUnitChar = 2
+		adjustToPt = 12.0
+	} else { // default is pixel
+		removeUnitChar = 0
+		adjustToPt = 1.0 / 96.0
+	}
```
Comment (sbinet) [inline svgbasic.go:231]: Go "good practices" recommend to use a `switch`.
```suggestion
	switch {
	case strings.HasSuffix(val, "pt"):
		removeUnitChar = 2
		adjustToPt = 1.0
	case strings.HasSuffix(val, "in"):
		removeUnitChar = 2
		adjustToPt = 72.0
	case strings.HasSuffix(val, "mm"):
		removeUnitChar = 2
		adjustToPt = 72.0 / 25.4
	case strings.HasSuffix(val, "cm"):
		removeUnitChar = 2
		adjustToPt = 72.0 / 2.54
	case strings.HasSuffix(val, "pc"):
		removeUnitChar = 2
		adjustToPt = 12.0
	default: // default is pixel
		removeUnitChar = 0
		adjustToPt = 1.0 / 96.0
	}
```

perhaps also add a case for `"px"` pixel units ?

File: svgwrite.go [hunk 3]
```diff
@@ -22,7 +22,8 @@ package fpdf
 
 // SVGBasicWrite renders the paths encoded in the basic SVG image specified by
 // sb. The scale value is used to convert the coordinates in the path to the
-// unit of measure specified in New(). The current position (as set with a call
+// unit of measure specified in New(). If scale its 0, automatically adapt SVG
```
Comment (sbinet) [inline svgwrite.go:25]:
```suggestion
// unit of measure specified in New(). If scale is 0, SVGBasicWrite automatically adapts the SVG document
```

File: svgwrite.go [hunk 4]
```diff
@@ -34,6 +35,9 @@ func (f *Fpdf) SVGBasicWrite(sb *SVGBasicType, scale float64) {
 	var path []SVGBasicSegmentType
 	var seg SVGBasicSegmentType
 	var startX, startY float64
+	if scale == 0.0 {
+		scale = 1.0 / f.k
```
Comment (sbinet) [inline svgwrite.go:39]: could `f.k` be 0 itself ?

File: svgwrite.go [hunk 5]
```diff
@@ -87,3 +91,68 @@ func (f *Fpdf) SVGBasicWrite(sb *SVGBasicType, scale float64) {
 		}
 	}
 }
+
+// Does same action than SVGBasicWrite(), but each SVG form are written
+// as path, so each path can be filled.
```
Comment (sbinet) [inline svgwrite.go:96]:
```suggestion
// SVGBasicDraw renders the paths in the provided SVGBasicType, but each SVG shape is written
// as a path that can be filled.
```

File: svgwrite.go [hunk 6]
```diff
@@ -87,3 +91,68 @@ func (f *Fpdf) SVGBasicWrite(sb *SVGBasicType, scale float64) {
...
+// the ellipse's perimeter. Filling uses the current fill color.
+func (f *Fpdf) SVGBasicDraw(sb *SVGBasicType, scale float64, styleStr string) {
+	originX, originY := f.GetXY()
+	var newX, newY float64
+	var cx0, cy0, cx1, cy1 float64
+	var path []SVGBasicSegmentType
+	var seg SVGBasicSegmentType
+	var startX, startY float64
+	if scale == 0.0 {
+		scale = 1.0 / f.k
```
Comment (sbinet) [inline svgwrite.go:110]: could `f.k` be 0 itself ?

Comment (thierry-f-78) [reply svgwrite.go:110]: No, in the fpdf.go file, the function which init doc "fpdfNew()" executes this (fpdf.go, line 126)

```go
 126    switch unitStr {
 127    case "pt", "point":
 128       f.k = 1.0
 129    case "mm":
 130       f.k = 72.0 / 25.4
 131    case "cm":
 132       f.k = 72.0 / 2.54
 133    case "in", "inch":
 134       f.k = 72.0
 135    default:
 136       f.err = fmt.Errorf("incorrect unit %s", unitStr)
 137       return
 138    }

```

So, `f.k` is set, or the fpdfNew() returns an error. 

however, the library may be initialized in the wrong way (eg: `f = &Fpdf{}`). In this case, `f.k` will be 0, but nothing will work anyway. I think a panic is acceptable in this situation.

File: svgwrite.go (hunk 4 above)
Comment (thierry-f-78) [reply svgwrite.go:39]: see below, same explanation.

**Review by sbinet — APPROVED** [review]
LGTM

thanks for the PR (and sticking with the somewhat lengthy review)

**Comment by thierry-f-78** [comment]
Thank for the code review and for maintaining this libray.

**Pull request merged**